### PR TITLE
interchange: enable clusters support

### PR DESCRIPTION
**Pull request by acomodi**
This PR is based on the ideas of https://github.com/YosysHQ/nextpnr/pull/703, and aims at providing an implementation of the clustering support, so that the placer can be guided by information defined in the chip info.

The current state of developement, combined with https://github.com/SymbiFlow/python-fpga-interchange/pull/91, is able to get a working counter test that uses carry-chains.

TODO:
- [x] Test murax design: currently takes quite a while to place if carry-chains are enabled.
- [x] Add LUT-to-FF clustering support to speed up placement
- [x] Iterate and refine the cluster chipinfo defintions to be more compact and cleaner
- [ ] Add comments

NOTE:
- The murax test design is not meant to be checked in, but is only useful in the various developement iterations.

### CODE REVIEW

File: fpga_interchange/arch.h
Comment (gatecat) [inline fpga_interchange/arch.h:708]: as a note, now that #718 has just been merged, this will need to be updated to use `dict` instead of `std::unordered_map`

File: fpga_interchange/arch_pack_clusters.cc
Comment (gatecat) [inline fpga_interchange/arch_pack_clusters.cc:206]: please use `||` not `or`

File: fpga_interchange/arch_pack_clusters.cc
Comment (gatecat) [inline fpga_interchange/arch_pack_clusters.cc:167]: I think using `pool` (i.e.  the new `unordered_set` replacement) would make more sense than `std::vector`, as far as I can tell the only thing you do with this structure is check membership (which can be done significantly faster with a hash set than `O(n)` in a vector) with nothing that cares about ordering

File: fpga_interchange/arch_pack_clusters.cc
Comment (gatecat) [inline fpga_interchange/arch_pack_clusters.cc:171]: I think `emplace_back` should work and is cleaner than `push_back(std::make_pair...`

File: fpga_interchange/arch_pack_clusters.cc
Comment (gatecat) [inline fpga_interchange/arch_pack_clusters.cc:250]: this hardcoded `CO3` needs either to be fixed, or a `FIXME` comment left explaining why it's there

File: fpga_interchange/arch_pack_clusters.cc
Comment (gatecat) [inline fpga_interchange/arch_pack_clusters.cc:248]: what if `next_net` has more than one user?

File: fpga_interchange/arch_pack_clusters.cc
Comment (gatecat) [inline fpga_interchange/arch_pack_clusters.cc:265]: these functions (this and `getClusterOffset`) are only intended to return approximations, but it's important that they function **before** the chain has been placed (that is in fact their main use, after the chain has been placed anything downstream that wants to know approximate bounds/offsets of the chain can just query actual positions). So this current implementation isn't really the intended purpose of these functions.

File: fpga_interchange/arch_pack_clusters.cc
Comment (gatecat) [inline fpga_interchange/arch_pack_clusters.cc:348]: another note post #718, this will need to be changed to `for (auto &cell : ctx->cells) {` `CellInfo *ci = cell.second.get();`

File: fpga_interchange/arch_pack_clusters.cc
Comment (gatecat) [inline fpga_interchange/arch_pack_clusters.cc:387]: can you give a bit more context behind this `disconnect_port` ?

Comment (acomodi) [reply fpga_interchange/arch_pack_clusters.cc]: Sure, comment added

File: fpga_interchange/arch_pack_clusters.cc
Comment (acomodi) [reply fpga_interchange/arch_pack_clusters.cc]: Fixed

File: fpga_interchange/arch_pack_clusters.cc
Comment (acomodi) [reply fpga_interchange/arch_pack_clusters.cc]: I might have missed something, but I have noticed that the `getClusterBounds` is not called at the moment in any of the algos. At the moment I left it blank, with a  `TODO`, to fill it as soon as it will be necessary.

I have also implemented a "fixed" heuristic for the clusters that span multiple tiles, hacked for the moment to return a valid offset for the carry-chains. I think that this kind of information could be provided as a device configuration in the cluster info.

File: fpga_interchange/arch_pack_clusters.cc
Comment (acomodi) [reply fpga_interchange/arch_pack_clusters.cc]: This has been reworked and the discovery of the chained cells is only done when preparing clusters and all the users are checked when doing so.

File: fpga_interchange/arch_pack_clusters.cc
Comment (acomodi) [reply fpga_interchange/arch_pack_clusters.cc]: this was actually a leftover, this information now comes from the chip info.

File: fpga_interchange/arch_pack_clusters.cc
Comment (acomodi) [reply fpga_interchange/arch_pack_clusters.cc]: Fixed

File: fpga_interchange/arch_pack_clusters.cc
Comment (acomodi) [reply fpga_interchange/arch_pack_clusters.cc]: Sure, I guess I just came back from python when writing this :)

File: fpga_interchange/arch.h
Comment (acomodi) [reply fpga_interchange/arch.h]: Changed all instances into `pool/dict`

**Comment by acomodi** [comment]
@gatecat I have reworked quite a bit the clustering approach for the interchange. With the current implementation I can get to a working murax design placed and routed with carry chains in ~8-9 minutes, where most of the time is still spent during placement.

For what I understand, there is still quite some time spent by the placer to place FFs which are not legal due to different S/R/CE nets w.r.t. already placed FFs in the same site.
One possible solution I though of, that might still require an extension of the clustering approach, is to have a sort of attraction function that gathers and organizes sets of FFs (or other) to be placed at the same site, thus making the whole legalization step quite faster, as all the compatible cells can be placed at once in the same site. Can this be a valid solution?

**Comment by gatecat** [comment]
>  One possible solution I though of, that might still require an extension of the clustering approach, is to have a sort of attraction function that gathers and organizes sets of FFs (or other) to be placed at the same site, thus making the whole legalization step quite faster, as all the compatible cells can be placed at once in the same site. Can this be a valid solution?

Yes and this seems reasonable in terms of the long-term direction for nextpnr, however it will need some discussion design wise to make sure it's flexible enough.

While I'm sure it will help a lot, I still think the biggest root-cause here is simply the validity check being 10-100x slower in interchange than other arches and speeding that up will be the easiest way to significantly improve placer runtime.

File: fpga_interchange/arch_pack_clusters.cc
Comment (gatecat) [reply fpga_interchange/arch_pack_clusters.cc]: yeah, there are some possible future placer ideas I had that meant it could be useful, but a stub is alright for now `getClusterOffset` is the more important one for HeAP.

**Comment by acomodi** [comment]
Ok, I'd say I may start looking into how to get some more performance out of the validity checks once we land the first clustering handling implementation.

Apart from the `getClusterOffset` implementation that is only effective for CARRY4, I believe this PR is now ready for another review.

File: fpga_interchange/arch_pack_clusters.cc
Comment (gatecat) [inline fpga_interchange/arch_pack_clusters.cc:242]: this should be removed, or gated behind one of the debug flags and/or a macro

File: fpga_interchange/arch_pack_clusters.cc
Comment (gatecat) [inline fpga_interchange/arch_pack_clusters.cc:341]: I think this should be `static`

File: fpga_interchange/arch_pack_clusters.cc
Comment (gatecat) [inline fpga_interchange/arch_pack_clusters.cc:429]: it's notable that "CIN connected to 0" and "CIN connected to 1" are cases that need to be told apart, both being supported but with distinct FASM features: https://github.com/SymbiFlow/prjxray-db/blob/c28433840752ff0da5b8353bfdeafed24148aa38/artix7/segbits_clbll_l.db#L674-L675

I guess the  plan is to determine this from the logical netlist during FASM generation, as the connectivity in the physical netlist will be lost?

Comment (acomodi) [reply fpga_interchange/arch_pack_clusters.cc:429]: I think this is an interesting case where the CIN and CYINIT BEL inputs are actually physically the same, but their connectivity is expressed differently for P&R purposes. 

If the `CYINIT` cell port is emitted by yosys, than there would be no problems and the port will not get disconnected, because the `PRECYINIT` routing BEL is connected to the general routing. The problem is that the `GND` net is assigned to the `CI` cell port as well, and this cell port is mapped to the `CIN` bel port, which does not have connections to `GND` (at least from a P&R perspective).

The root of a chain, all the `C0`, `C1`, `AX` connections can be present for `CYINIT`, and the information preserved in the physical netlist as well.

Comment (gatecat) [reply fpga_interchange/arch_pack_clusters.cc:438]: Yeah, that seems reasonable. We might need something somewhere to move constants from `CIN` to `CYINIT` if this proves to be a problem.

File: fpga_interchange/arch_pack_clusters.cc
Comment (acomodi) [reply fpga_interchange/arch_pack_clusters.cc]: This was actually some debug prints as CI is currently failing and I wanted to get some more data on why. Removed this log

File: fpga_interchange/arch_pack_clusters.cc
Comment (acomodi) [reply fpga_interchange/arch_pack_clusters.cc]: Right, fixed

**Comment by acomodi** [comment]
After some debugging, the CI failure is caused by [this issue](https://github.com/Xilinx/RapidWright/issues/183). The pin mappings for the CARRY4 are incomplete and should get added in the next RW release.

For the time being I disabled carry inference, and the only cluster that will be applied is going to be the `LUT+FF` one.

**Comment by gatecat** [comment]
We could always modify CI to download that patch

**Comment by acomodi** [comment]
> We could always modify CI to download that patch

Yes, this is actually better so we make sure nothing is broken with BELs chaining

**Review by gatecat — APPROVED** [review]
LGTM, I'll merge once the CI has gone green

**Pull request merged**
